lib/pharos/phases/configure_host.rb

```diff
@@ -30,8 +32,58 @@ def call
       end
 
       def configure_container_runtime
-        logger.info "Configuring container runtime (#{@host.container_runtime}) packages ..."
-        host_configurer.configure_container_runtime
+        if @host.new? || host_configurer.configure_container_runtime_safe?
+          logger.info "Configuring container runtime (#{@host.container_runtime}) packages ..."
+          host_configurer.configure_container_runtime!
+        else
+          throttled_work('reconfigure_with_drain', rollout_concurrency) do
+            logger.info "Reconfiguration of container runtime (#{@host.container_runtime}) might affect workloads, switching to a safe mode ..."
+            reconfigure_with_drain
+          end
+        end
+      end
+
+      # @return [Integer]
+      def rollout_concurrency
+        (@config.hosts.size * ROLLOUT_PERCENTAGE / 100).ceil
+      end
+
+      def reconfigure_with_drain
+        if master_healthy?
+          logger.info "Draining node ..."
+          begin
+            drain_host
+          rescue Pharos::ExecError
+            drain_host!
+          end
+        else
+          logger.warn "Cannot drain node because control-plane is not healthy ..."
+        end
+
+        logger.info "Waiting for volume umounts ..."
+        sleep 1 until @host.transport.exec("sudo mount | grep /var/lib/kubelet/plugins").error?
+
+        logger.info "Reconfiguring container runtime (#{@host.container_runtime}) packages ..."
+        host_configurer.configure_container_runtime!
+
+        return unless master_healthy?
+
+        logger.info "Uncordoning node ..."
+        sleep 1 until master_host.transport.exec("kubectl uncordon #{@host.hostname}").success?
+        logger.info "Waiting for node to be ready ..."
+        sleep 10 until master_host.transport.exec("kubectl get nodes -o jsonpath=\"{range .items[*]}{@.metadata.name}:{range @.status.conditions[*]}{@.type}={@.status};{end}{end}\" | grep 'Ready=True'").success?
```

Comment:
I don't see how this picks up the correct node.

```
$ kubectl get nodes -o jsonpath="{range .items[*]}{@.metadata.name}:{range @.status.conditions[*]}{@.type}={@.status};{end}{end}"
host-00:NetworkUnavailable=False;MemoryPressure=False;DiskPressure=False;PIDPressure=False;Ready=True;InvalidLicense=False;host-01:NetworkUnavailable=False;MemoryPressure=False;DiskPressure=False;PIDPressure=False;Ready=True;host-02:NetworkUnavailable=False;MemoryPressure=False;DiskPressure=False;PIDPressure=False;Ready=True;
```

To me, it appears this will be true when there is at least one node with `Ready=True`.